<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="John Blischak" />

<meta name="date" content="2018-08-20" />

<title>The benefit of sharing information across genes</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">DE analysis with limma</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jdblischak/dc-bioc-limma">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">The benefit of sharing information across genes</h1>
<h4 class="author"><em>John Blischak</em></h4>
<h4 class="date"><em>2018-08-20</em></h4>

</div>


<p><strong>Last updated:</strong> 2018-08-20</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(12345)</code> </summary></p>
<p>The command <code>set.seed(12345)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<p><details> <summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/jdblischak/dc-bioc-limma/tree/1526cb167a458b3d5d09f176d7ccfb644a534d67" target="_blank">1526cb1</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  code/counts_per_sample.txt
    Untracked:  code/table-s1.txt
    Untracked:  docs/figure/linear-models.Rmd/

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes. </details>
</li>
</ul>
<details> <summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/jdblischak/dc-bioc-limma/blob/192083341b74a84f8295ca7bf85f48e443865751/analysis/sharing.Rmd" target="_blank">1920833</a>
</td>
<td style="text-align:left;">
John Blischak
</td>
<td style="text-align:left;">
2018-08-20
</td>
<td style="text-align:left;">
Refactor first edition of chapter 1 into distinct lessons.
</td>
</tr>
</tbody>
</table>
</ul>
<p></details></p>
<hr />
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The simulation and visualizations below demonsrate the differences in the results due to limma sharing information across genes to shrink the estimates of the variance.</p>
</div>
<div id="setup" class="section level2">
<h2>Setup</h2>
<pre class="r"><code>library(&quot;cowplot&quot;)
library(&quot;dplyr&quot;)
library(&quot;ggplot2&quot;)
theme_set(theme_classic(base_size = 16))
library(&quot;knitr&quot;)
opts_chunk$set(fig.width = 10, fig.height = 5, message = FALSE)
library(&quot;stringr&quot;)
library(&quot;tidyr&quot;)</code></pre>
</div>
<div id="simulation" class="section level2">
<h2>Simulation</h2>
<p>Create some synthetic data for illustrating concepts. The simulated gene expression matrix has 100 genes and 6 samples (3 treatment and 3 control).</p>
<pre class="r"><code>set.seed(12345)
create_exp_mat &lt;- function(n1, n2, ng,
                           alpha_mean, beta_mean, epsilon_sd) {
  status &lt;- c(rep(0, n1), rep(1, n2))
  ns &lt;- length(status)
  status &lt;- matrix(status, nrow = 1)

  alpha &lt;- rnorm(ng, mean = alpha_mean, sd = 1)
  beta &lt;- matrix(rnorm(ng, mean = beta_mean, sd = 1), ncol = 1)
  epsilon &lt;- matrix(rnorm(ng * ns, mean = 0, sd = epsilon_sd),
                    nrow = ng, ncol = ns)
  Yg &lt;- alpha + beta %*% status + epsilon
  return(Yg)
}

gexp &lt;- rbind(
  # 30 non-DE genes with high variance
  create_exp_mat(n1 = 3, n2 = 3, ng = 30, alpha_mean = 10, beta_mean = -1:1, epsilon_sd = 3),
  # 30 non-DE genes with low variance
  create_exp_mat(n1 = 3, n2 = 3, ng = 30, alpha_mean = 10, beta_mean = -1:1, epsilon_sd = 1),
  # 10 upregulated DE genes with low variance
  create_exp_mat(n1 = 3, n2 = 3, ng = 10, alpha_mean = 10, beta_mean = 5, epsilon_sd = 1),
  # 10 upregulated DE genes with high variance
  create_exp_mat(n1 = 3, n2 = 3, ng = 10, alpha_mean = 10, beta_mean = 5, epsilon_sd = 3),
  # 10 downregulated DE genes with low variance
  create_exp_mat(n1 = 3, n2 = 3, ng = 10, alpha_mean = 10, beta_mean = -5, epsilon_sd = 1),
  # 10 downregulated DE genes with high variance
  create_exp_mat(n1 = 3, n2 = 3, ng = 10, alpha_mean = 10, beta_mean = -5, epsilon_sd = 3)
)

# Add names for samples
group &lt;- rep(c(&quot;con&quot;, &quot;treat&quot;), each = ncol(gexp) / 2)
samples &lt;- paste0(group, 1:3)
colnames(gexp) &lt;- samples

# Add names for genes
genes &lt;- sprintf(&quot;gene%02d&quot;, 1:nrow(gexp))
rownames(gexp) &lt;- genes

heatmap(gexp)</code></pre>
<p><img src="figure/sharing.Rmd/simulation-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="standard-linear-model" class="section level2">
<h2>Standard linear model</h2>
<p>Find differentially expressed genes using a standard linear model.</p>
<pre class="r"><code>lm_beta &lt;- numeric(length = nrow(gexp))
lm_se &lt;- numeric(length = nrow(gexp))
lm_p &lt;- numeric(length = nrow(gexp))
for (i in 1:length(lm_p)) {
  mod &lt;- lm(gexp[i, ] ~ group)
  result &lt;- summary(mod)
  lm_beta[i] &lt;- result$coefficients[2, 1]
  lm_se[i] &lt;- result$coefficients[2, 2]
  lm_p[i] &lt;- result$coefficients[2, 4]
}
hist(lm_p, xlab = &quot;p-values&quot;, main = &quot;Standard linear model&quot;)</code></pre>
<p><img src="figure/sharing.Rmd/lm-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="limma-linear-model" class="section level2">
<h2>limma linear model</h2>
<p>Find differentially expressed genes using limma.</p>
<pre class="r"><code>library(&quot;limma&quot;)
design &lt;- model.matrix(~group)
colnames(design) &lt;- c(&quot;Intercept&quot;, &quot;treat&quot;)
fit &lt;- lmFit(gexp, design)
head(fit$coefficients)</code></pre>
<pre><code>       Intercept      treat
gene01 11.316083 -2.4577980
gene02  9.833304  2.7130980
gene03 12.653098 -0.2048963
gene04 12.275601  0.2934781
gene05  8.617135  2.3383110
gene06  5.878178  3.9361382</code></pre>
<pre class="r"><code>fit &lt;- eBayes(fit)
results &lt;- decideTests(fit[, 2])
summary(results)</code></pre>
<pre><code>       treat
Down      15
NotSig    71
Up        14</code></pre>
<pre class="r"><code>stats &lt;- topTable(fit, coef = &quot;treat&quot;, number = nrow(fit), sort.by = &quot;none&quot;)
hist(stats[, &quot;P.Value&quot;], xlab = &quot;p-values&quot;, main = &quot;limma linear model&quot;)</code></pre>
<p><img src="figure/sharing.Rmd/limma-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="comparison" class="section level2">
<h2>Comparison</h2>
<p>Compare the p-values from <code>lm</code> and <code>limma</code> (both adjusted for multiple testing with the BH FDR).</p>
<pre class="r"><code>stats &lt;- cbind(stats,
               sd = apply(gexp, 1, sd),
               var = apply(gexp, 1, var),
               lm_beta, lm_se,
               lm_p = p.adjust(lm_p, method = &quot;BH&quot;))

stats$labels_pre &lt;- c(rep(&quot;non-DE; high-var&quot;, 30),
                      rep(&quot;non-DE; low-var&quot;, 30),
                      rep(&quot;DE-up; low-var&quot;, 10),
                      rep(&quot;DE-up; high-var&quot;, 10),
                      rep(&quot;DE-down; low-var&quot;, 10),
                      rep(&quot;DE-down; high-var&quot;, 10))

stats$labels &lt;- rep(&quot;non-DE&quot;, nrow(stats))
stats$labels[stats$adj.P.Val &lt; 0.05 &amp; stats$lm_p &lt; 0.05] &lt;- &quot;DE&quot;
stats$labels[stats$adj.P.Val &lt; 0.05 &amp; stats$lm_p &gt;= 0.05] &lt;- &quot;limma-only&quot;
stats$labels[stats$adj.P.Val &gt;= 0.05 &amp; stats$lm_p &lt; 0.05] &lt;- &quot;lm-only&quot;
table(stats$labels)</code></pre>
<pre><code>
        DE limma-only    lm-only     non-DE 
        22          7          3         68 </code></pre>
<pre class="r"><code>table(stats$labels, stats$labels_pre)</code></pre>
<pre><code>            
             DE-down; high-var DE-down; low-var DE-up; high-var
  DE                         0                7               3
  limma-only                 1                3               1
  lm-only                    0                0               0
  non-DE                     9                0               6
            
             DE-up; low-var non-DE; high-var non-DE; low-var
  DE                      9                0               3
  limma-only              1                1               0
  lm-only                 0                0               3
  non-DE                  0               29              24</code></pre>
<pre class="r"><code>stopifnot(stats$logFC == stats$lm_beta)

de &lt;- data.frame(effect_size = stats$lm_beta,
                 std_dev = stats$sd,
                 lm = stats$lm_p &lt; 0.05,
                 limma = stats$adj.P.Val &lt; 0.05)

head(de)</code></pre>
<pre><code>  effect_size  std_dev    lm limma
1  -2.4577980 3.750555 FALSE FALSE
2   2.7130980 3.353835 FALSE FALSE
3  -0.2048963 1.715542 FALSE FALSE
4   0.2934781 3.511502 FALSE FALSE
5   2.3383110 1.955378 FALSE FALSE
6   3.9361382 3.326517 FALSE FALSE</code></pre>
<pre class="r"><code># View the number of discrepancies
table(de$lm, de$limma)</code></pre>
<pre><code>       
        FALSE TRUE
  FALSE    68    7
  TRUE      3   22</code></pre>
<pre class="r"><code># Plot effect size (y-axis) vs. standard deviation (x-axis)
ggplot(de, aes(x = std_dev, y = effect_size, color = limma)) +
  geom_point()</code></pre>
<p><img src="figure/sharing.Rmd/comparison-1.png" width="960" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggplot(stats, aes(x = sd, y = logFC, color = labels)) +
  geom_point()</code></pre>
<p><img src="figure/sharing.Rmd/comparison-2.png" width="960" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggplot(stats, aes(x = logFC, y = -log10(P.Value), color = labels)) +
  geom_point()</code></pre>
<p><img src="figure/sharing.Rmd/comparison-3.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="example-genes" class="section level2">
<h2>Example genes</h2>
<p>Visualize example genes with boxplots. Note that the limma-only gene has higher variance compared to the lm-only gene.</p>
<pre class="r"><code># Find a good example of a DE gene
index &lt;- which(stats$labels_pre == &quot;DE-up; low-var&quot; &amp; stats$labels == &quot;DE&quot;)[1]
single_gene &lt;- gexp %&gt;% as.data.frame %&gt;%
  slice(index) %&gt;%
  gather(key = &quot;group&quot;, value = &quot;gene&quot;) %&gt;%
  mutate(group = str_extract(group, &quot;[a-z]*&quot;)) %&gt;%
  as.data.frame()

# Find a gene that is DE for both, DE for lm-only, and DE for limma-only
de_not &lt;- de_lm &lt;- which(stats$labels == &quot;non-DE&quot; &amp;
                           stats$labels_pre == &quot;non-DE; high-var&quot; &amp;
                           stats$logFC &gt; 0)[1]
de_both &lt;- which(stats$labels == &quot;DE&quot; &amp;
                   stats$labels_pre == &quot;DE-up; low-var&quot;)[1]
de_lm &lt;- which(stats$labels == &quot;lm-only&quot; &amp;
                 stats$labels_pre == &quot;non-DE; low-var&quot; &amp;
                 stats$logFC &gt; 0)[1]
de_limma &lt;- which(stats$labels == &quot;limma-only&quot; &amp;
                    stats$labels_pre == &quot;DE-up; high-var&quot;)[1]

compare &lt;- gexp %&gt;%
  as.data.frame() %&gt;%
  slice(c(de_not, de_both, de_lm, de_limma)) %&gt;%
  mutate(type = c(&quot;neither&quot;, &quot;both&quot;, &quot;lm-only&quot;, &quot;limma-only&quot;)) %&gt;%
  gather(key = &quot;group&quot;, value = &quot;gene&quot;, con1:treat3) %&gt;%
  mutate(group = str_extract(group, &quot;[a-z]*&quot;)) %&gt;%
  as.data.frame()

head(compare)</code></pre>
<pre><code>        type group      gene
1    neither   con  6.681872
2       both   con  8.555641
3    lm-only   con  9.959914
4 limma-only   con 11.391149
5    neither   con  8.144218
6       both   con  7.977472</code></pre>
<pre class="r"><code># Plot gene expression (gene; y-axis) vs. group (x-axis)
ggplot(compare, aes(x = group, y = gene)) +
  geom_boxplot() +
  facet_wrap(~type, nrow = 1)</code></pre>
<p><img src="figure/sharing.Rmd/example-genes-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.0 (2018-04-23)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 17134)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] bindrcpp_0.2.2 limma_3.36.1   tidyr_0.8.1    stringr_1.3.1 
[5] knitr_1.20     dplyr_0.7.5    cowplot_0.9.2  ggplot2_2.2.1 

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.17      compiler_3.5.0    pillar_1.2.3     
 [4] git2r_0.21.0      plyr_1.8.4        workflowr_1.1.1  
 [7] bindr_0.1.1       R.methodsS3_1.7.1 R.utils_2.6.0    
[10] tools_3.5.0       digest_0.6.15     evaluate_0.10.1  
[13] tibble_1.4.2      gtable_0.2.0      pkgconfig_2.0.1  
[16] rlang_0.2.1       yaml_2.1.19       rprojroot_1.3-2  
[19] grid_3.5.0        tidyselect_0.2.4  glue_1.2.0       
[22] R6_2.2.2          rmarkdown_1.10    purrr_0.2.5      
[25] magrittr_1.5      whisker_0.3-2     backports_1.1.2  
[28] scales_0.5.0      htmltools_0.3.6   assertthat_0.2.0 
[31] colorspace_1.3-2  labeling_0.3      stringi_1.2.3    
[34] lazyeval_0.2.1    munsell_0.5.0     R.oo_1.22.0      </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.1.1
</p>
<hr>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
